### DISCUSSION TITLE
JIT and browser based CSS building

### DISCUSSION THREAD

**Discussion by leojpod** [tailwindlabs/tailwindcss#6491] [Help] [answered]
Hello! 

I've been working with tailwind since the v1 came around and overall that's the best CSS experience I've got. 
My usual UI-building goes like this: 

- make a draft of what I want in my code (usually Elm or Clojurescript)
- open the developer toolbox on my browser (firefox or chrome-based variant)
- start messing around with the html directly from the toolbox pane until I find the combination of classes I like to do the job I want
- go back to the code and apply that

I have several reasons to have this workflow: 

- no time lost in rebuilding the code to see what it would look like with a different class
- if the piece of UI I'm working on is a confirm screen for instance or something that requires user action to reach I might loose that state upon hot-reloading and it breaks the flow of my thoughts/ideas
- it's the fastest way I've found to have quick feedback

The problem is that with V3 and the always-JIT is that my browser now knows only the classes I've already used in the project which means that even if I give up on the autocompletion from the browser, typing `bg-blue-400` will only work in my browser if that class is already used elsewhere in the project. 

After a bit of research, I found that there is no way to turn of JIT. The best way I've found is to use something like https://www.npmjs.com/package/tailwind-safelist-generator (which sadly won't install on a v3 project) and build a `safelist.txt` with everything when I'm developing. 

Is there something better? I'd love a magic thing I can wave to get all the classes from time to time, in the meantime I'll generate a big `safelist` :) 

P.S.: this post might come across as a negative one, but it's not. I really enjoy working with tailwind, and it's the perfect partner for functional programming on the FE. Even with that mild inconvenient, I don't know of a better alternative.

**Comment by leojpod** [comment]
Edit: 
The simplest way I found was to misuse the [`safelist`](https://tailwindcss.com/docs/content-configuration#safelisting-classes) config with a pattern like this: 

```javascript
  safelist: [
    {
      pattern: /.*/,
    },
  ],
```

**Reply by leojpod** [reply]
can always wrap that in a js function that detect the environment and do something with it.

**Comment by NovusOpus** [comment]
Does the [Tailwind Play CDN](https://tailwindcss.com/docs/installation/play-cdn) works for your purpose?

Seems like you're going on the right path with the Safelist Patterns, Please if you manage to figure out, share your findings since many users where looking for a way to extract or generate a CSS File with all Utilities/Classes.

**Comment by philwolstenholme** [comment]
Keep an eye on Marcel Pociot's work on some custom Tailwind devtools via browser extensions: https://twitter.com/marcelpociot/status/1470442562561396737

**Reply by leojpod** [reply]
that sounds rather nice!

**Comment by leojpod** [comment, marked as answer]
so what I have so far is this: 

```javascript
  safelist: process.env.SAFELISTING
    ? [
        {
          pattern: /.*/,
          variants: ["hover", "focus", "dark"],
        },
      ]
    : [],
```

which is coupled with tailwind being called like this in my `package.json`: 

```sh
SAFELISTING=true npm run css:build -- --watch
```

few things that might be interesting for other people's setup: 

- I don 't think you can pattern-match the variants
- I have put only a few variants in there 'cause my main usage is mostly to identify which classes I'm after and then I apply the variants that make sense (for responsive content, hover, focus, group-stuff, ... )
